### PR TITLE
nixos/nextcloud: fix nginx routing to store and nix apps

### LINKED ISSUE #277206: nextcloud28: Forms page is blank
### Describe the bug
The extraApp called ``forms`` displays as blank.  
This was working fine on nextcloud27.  
Rest of the site is working fine.

### Steps To Reproduce

1. I am currently on ``'github:NixOS/nixpkgs/5f64a12a728902226210bf01d25ec6cbb9d9265b' (2023-12-24)``
2. Have ``forms`` in ``extraApps``
3. Go to webpage
4. It is blank on:
   * Firefox
   * Chrome
   * Edge

### Expected behavior
It should display the normal forms page.

### Screenshots
image

### Additional context
Found non-nixos issue relating to it here: 

* https://github.com/nextcloud/forms/issues/1838
* https://github.com/linuxserver/docker-nextcloud/issues/388

Seems it may be a ``content-type`` related issue?

### Notify maintainers
@schneefux @bachp @globin @ma27

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.5.11-6-pve, NixOS, 24.05 (Uakari), 24.05pre-git`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - nixpkgs: `not found`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by Ma27**
## Description of changes
Closes #277206

The bug mentioned above was a symptom of the issue fixed here: when opening the `forms` app which is installed via `extraApps` (or the app store) the site wouldn't work because `.mjs` files had the wrong Content-Type.

The actual problem got fixed already[1], however this config was not used for stuff from `/nix-apps` & `/store-apps` which had their own location section with only a `root ;` statement.

In fact, this setup isn't strictly supported by Nextcloud upstream[2], so to fix this for good, I decided to follow the upstream suggestion for app directories outside the server root, i.e. linking them back into the store path.

This means that the module generates a new derivation now with

* `services.nextcloud.package` linked into it via `lndir`.
* under `nix-apps` is a symlink to the link farm containing all apps from `services.nextcloud.extraApps`.
* under `store-apps` is a symlink to `/var/lib/nextcloud/store-apps`. Since this is only used in the NixOS module that also configures this location for imperatively installed apps, this seems an OK thing to do.

Successfully tested the change on a productive Nextcloud 28.0.1 with several apps installed via `extraApps` (`forms`, `cospend`, `maps`, `user_saml` and a few more).

[1] 292c74c7a92efe45cfdfc091fb48d8c575346a95
[2] https://docs.nextcloud.com/server/28/admin_manual/apps_management.html#using-custom-app-directories

cc @Silver-Golden for testing.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Silver-Golden** [comment]
I'lltest this out this evening and get back to ye.    
Thanks for the prompt response.

**Comment by SuperSandro2000** [comment]
I don't think we really want to backport this to 23.05 which supports ends literally in the next days.

File: nixos/modules/services/web-apps/nextcloud.nix [hunk 1]
```diff
@@ -23,6 +23,45 @@ let
...
+    };
+  };
+
+  webroot = pkgs.runCommand
+    "${cfg.package.name or "nextcloud"}-with-apps"
+    { nativeBuildInputs = [ pkgs.xorg.lndir ]; }
+    ''
+      mkdir $out
+      pushd "$out" &>/dev/null
+        lndir "${cfg.package}"
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/nextcloud.nix:53]: lndir also supports a target directory, so we could avoid the pushd/popd.

Also do we really need to use lndir or would ln -s be enough?

**Review by SuperSandro2000 — APPROVED** [review]
I've cherry-picked the commit into my fork and deployed it to my nextcloud which only has store-apps, no apps installed via nix, and I didn't encounter any issues.

**Review by beardhatcode — COMMENTED** [review]
I no not have time to test this on my server atm, but it looks good

**Comment by Silver-Golden** [comment]
I copied the ``nextcloud.nix`` file locally, added ``disabledModules = [ "services/web-apps/nextcloud.nix" ];`` to disable the existing one and then imported that file.

Forms work as expected.

File: nixos/modules/services/web-apps/nextcloud.nix (hunk 1 above)
Comment (Ma27) [reply nixos/modules/services/web-apps/nextcloud.nix]: Good point, done with `ln -s`!

**Pull request merged**